DM-27169: Use FilterLabel in Exposure/ExposureInfo

kfindeisen opened the pull request:
This PR adds support for `FilterLabel` as a persisted form for filter information in decomposed exposures. Must be merged with lsst/afw#551. `Filter` remains both for backward-compatibility, and for consistency with the `Exposure` API (which temporarily also has `getFilter()` and `getFilterLabel()`). 

The plan for the `Exposure` class is to migrate code to using the `FilterLabel` class (through `getFilterLabel()`) on DM-27170, remove `Filter` and `getFilter()` and replace them with a new `getFilter()` that returns a `FilterLabel` on DM-27177, then remove `getFilterLabel()` on DM-27811. I propose that the components follow the same pattern to minimize user confusion.

kfindeisen committed:
Support persistence of FilterLabel.

The Exposure.filter component remains as Filter for API compatibility.
Filter information will be migrated to the FilterLabel class over
DM-27170 and DM-27177, then migrated back to the filter component
over DM-27177 and DM-27811.

timj commented on python/lsst/daf/butler/configs/datastores/formatters.yaml:
```diff
@@ -13,6 +13,7 @@ CoaddInputs: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
 VisitInfo: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
 ApCorr: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
 PhotoCalib: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
+FilterLabel: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
```
It seems to me that serializing in FITS is overkill. A FilterLabel is a couple of key/value pairs isn't it? YAML or JSON seems fine for that. We could fairly trivially make the FilterFormatter work for FilterLabel and Filter since Formatters know the pytype they are formatting.

kfindeisen replied:
Every component of `ExposureInfo`, now including `FilterLabel`, is held, managed, and persisted using the same framework; this makes the code much simpler and lets us avoid (or, at least, phase out) a lot of the special-case code of the old `ExposureInfo`. Given that `FilterLabel`/`ExposureInfo` already provides a FITS format in order to benefit from these advantages, why should we create a second format instead of reusing one that is supported out of the box? If the content of `FilterLabel` changes (which is quite possible, as the filter system evolves), should we need to remember to synchronize changes to two different representations?

In my view, the custom `FilterFormatter` was a necessary evil that standardization has made unnecessary.

timj reviewed (approved):
These changes are fine but I'd like to know the justification of writing FilterLabel as FITS.

kfindeisen committed:
Clarify derived disassembled component handling.

Add a link to Formatter to make that part of the documentation
more prominent, and add a note about supporting the delegation.
